diff --git a/app/src/main/java/com/yesmanchik/camerastreamer/ImageSender.java b/app/src/main/java/com/yesmanchik/camerastreamer/ImageSender.java
index 2dc41c5..2ac4466 100644
--- a/app/src/main/java/com/yesmanchik/camerastreamer/ImageSender.java
+++ b/app/src/main/java/com/yesmanchik/camerastreamer/ImageSender.java
@@ -30,6 +30,9 @@ public class ImageSender {
             writeInt(bytes.length);
             stream.write(bytes);
         } catch (IOException e) {
+            String h = host;
+            host = "";
+            connect(h, port);
         }
     }
 
